src/typechecker/types.lisp

```diff
@@ -305,7 +305,7 @@
   (declare (type ty ty)
            (type fixnum num))
 
-  (assert (>= num (length (function-type-arguments ty))))
+  (assert (<= num (length (function-type-arguments ty))))
```

Comment:
This would throw type errors like
```lisp
The assertion
(COMMON-LISP:>= COALTON-IMPL/TYPECHECKER/TYPES::NUM
                #1=(COMMON-LISP:LENGTH
                    (COALTON-IMPL/TYPECHECKER/TYPES:FUNCTION-TYPE-ARGUMENTS
                     COALTON-IMPL/TYPECHECKER/TYPES:TY)))
failed with COALTON-IMPL/TYPECHECKER/TYPES::NUM = 1, #1# = 2.
   [Condition of type COMMON-LISP:SIMPLE-ERROR]
```
on code like
```clojure
COALTON-USER> (coalton (let ((f (fn (x) (fn (y) (* x y))))) ((f 1) 2)))
```
or
```clojure
COALTON-USER> (coalton-toplevel (define (f x) (fn (x) (fn (y) (* x y)))))
```

Reply:
@amorphedstar can we add a test for this